java/vector/src/main/java/org/apache/arrow/vector/dictionary/BatchedDictionary.java

```diff
@@ -0,0 +1,270 @@
+import org.apache.arrow.vector.types.pojo.FieldType;
+
+/**
+ * A dictionary implementation that can be used when writing batches of data to
+ * a stream or file. Supports delta or replacement encoding.
+ */
+public class BatchedDictionary implements Closeable, BaseDictionary {
+
+  private final DictionaryEncoding encoding;
+
+  private final BaseVariableWidthVector dictionary;
+
+  private final BaseIntVector indexVector;
+
+  private final DictionaryHashTable hashTable;
+
+  private final boolean forFileIPC;
+
+  private int deltaIndex;
+
+  private int dictionaryIndex;
+
+  private boolean wasReset;
+
+  /**
+   * Creates a dictionary with two vectors of the given types. The dictionary vector
+   * will be named "{name}-dictionary".
+   * <p>
+   * To use this dictionary, provide the dictionary vector to a {@link DictionaryProvider},
+   * add the {@link #getIndexVector()} to the {@link org.apache.arrow.vector.VectorSchemaRoot}
+   * and call the {@link #setSafe(int, byte[], int, int)} or other set methods.
+   *
+   * @param name A name for the vector and dictionary.
+   * @param encoding The dictionary encoding to use.
+   * @param dictionaryType The type of the dictionary data.
+   * @param indexType The type of the encoded dictionary index.
+   * @param forFileIPC Whether the data will be written to a file or stream IPC. Throws an
+   *                   exception if a replacement dictionary is provided to a file IPC.
+   * @param allocator The allocator to use.
+   */
+  public BatchedDictionary(
+      String name,
+      DictionaryEncoding encoding,
+      ArrowType dictionaryType,
+      ArrowType indexType,
+      boolean forFileIPC,
+      BufferAllocator allocator
+  ) {
+    this(name, encoding, dictionaryType, indexType, forFileIPC, allocator, "-dictionary");
+  }
+
+  /**
+   * Creates a dictionary with two vectors of the given types.
+   *
+   * @param name A name for the vector and dictionary.
+   * @param encoding The dictionary encoding to use.
+   * @param dictionaryType The type of the dictionary data.
+   * @param indexType The type of the encoded dictionary index.
+   * @param forFileIPC Whether the data will be written to a file or stream IPC. Throws an
+   *                   exception if a replacement dictionary is provided to a file IPC.
+   * @param allocator The allocator to use.
+   * @param suffix A non-null suffix to append to the name of the dictionary.
+   */
+  public BatchedDictionary(
+      String name,
+      DictionaryEncoding encoding,
+      ArrowType dictionaryType,
+      ArrowType indexType,
+      boolean forFileIPC,
+      BufferAllocator allocator,
+      String suffix
+  ) {
+    this.encoding = encoding;
+    this.forFileIPC = forFileIPC;
+    FieldVector vector = new FieldType(false, dictionaryType, null)
+        .createNewSingleVector(name + suffix, allocator, null);
+    if (!(BaseVariableWidthVector.class.isAssignableFrom(vector.getClass()))) {
```

Comment:
It seems like we should be able to evaluate dictionaryType before creating the vector rather than validating the type constructed afterwards.

Reply:
Yeah, that works.